api/http-api.md

```diff
@@ -168,10 +170,471 @@ Current suggested approach, subject to discussion:
+  "workflowType": { "name": "MyWorkflow" },
+  "taskQueue": { "name": "my-task-queue" }
+}'
+```
+
+Response:
+
+```json
+{
+  "runId": "7124ede7-fef5-4e15-8411-baa23d6beefd"
+}
+```
+
+### Start a workflow with input
+
+Request:
+
+```bash
+curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id -d '{
+  "workflowType": { "name": "MyWorkflow" },
+  "taskQueue": { "name": "my-task-queue" },
+  "input": ["Hello, World!"]
+}'
+```
+
+This assumes that the approach from "Payload Formatting" above is applied and works.
+
+Response:
+
+```json
+{
+  "runId": "7124ede7-fef5-4e15-8411-baa23d6beefd"
+}
+```
+
+### Signal workflow
+
+Request:
+
+```bash
+curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id/signal/MySignal
+```
+
+Response:
+
+```json
+{}
+```
+
+### Query workflow
+
+Request:
+
+```bash
+curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id/query/MyQuery
+```
+
+Response:
+
+```json
+{
+  "queryResult": ["Hello, World!"]
+}
+```
+
+We could set `queryResult` to be the implied body, but it is best we stick with the gRPC protocol we've bad for better
+or worse. So `queryRejected` comes back as a success. This is just how we chose to implement query rejection (not to be
+confused with query failure).
+
+### Update workflow without ID
+
+Request:
+
+```bash
+curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id/update/MyUpdate -d '{
+  "request": {
+    "input": {
+      "args": ["Hello, World!"]
+    }
+  }
+}'
+```
+
+Response:
+
+```json
+{
+  "updateRef": {
+    "workflowExecution": {
+      "workflowId": "my-workflow-id",
+      "runId": "7124ede7-fef5-4e15-8411-baa23d6beefd"
+    },
+    "updateId": "6dad5439-40dd-4b01-9f15-f5723a450776"
+  },
+  "outcome": {
+    "success": ["Hello, World!"]
+  }
+}
+```
+
+### Update workflow with ID
+
+Request:
+​
+```bash
+curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id/updates/my-update-id -d '{
+  "request": {
+    "input": {
+      "name": "MyUpdate",
+      "args": ["Hello, World!"]
+    }
+  }
+}'
+```
+​
+Response:
+​
+```json
+{
+  "updateRef": {
+    "workflowExecution": {
+      "workflowId": "my-workflow-id",
+      "runId": "7124ede7-fef5-4e15-8411-baa23d6beefd"
+    },
+    "updateId": "my-update-id"
+  },
+  "outcome": {
+    "success": ["Hello, World!"]
+  }
+}
+```
+
+So `/update/<update-name>` is sending to a name, but `/updates/<update-id>` is also sending to a name but an ID url too.
+The reason for this is to support a stable ID-based URL for creating and describing (i.e. non-blocking-result fetch) one
+day.
+
+### Cancel workflow
+
+Request:
+​
+```bash
+curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id/cancel -d '{
+  "reason": "some reason"
+}'
+```
+​
+Response:
+​
+```json
+{ }
+```
+
+## Rejected Alternative Approach - Full API Redesign
```

Comment:
I think there is a "middle path" here that we haven't considered, which is to extend the gRPC API to be more RESTful (where needed; seems like it's already close?), make that an explicit design goal of gRPC going forward, and then build the HTTP API on top of that.

We still get the benefits of having HTTP be "merely" a layer on top of gRPC (reduced time to ship, easier to maintain parity between gRPC and HTTP), and we also get something that looks nice HTTP-wise.

And yes, I'm aware this is just shifting the problem around, code-wise, but I think it might also reduce some of the cons you mention below re. re-reviewing every single gRPC API, and being at the mercy of gRPC.

I'm not advocating we have to clean up gRPC right now, necessarily—just that we give it a bit of thought, capture pros/cons here, and see if this is something we can work towards.

Reply:
> I think there is a "middle path" here that we haven't considered, which is to extend the gRPC API to be more RESTful (where needed; seems like it's already close?), make that an explicit design goal of gRPC going forward, and then build the HTTP API on top of that.

It was never considered it because I don't think altering or adding more gRPC calls is acceptable for this project.

> seems like it's already close?

Can you give a concrete example here?

> And yes, I'm aware this is just shifting the problem around, code-wise

Not just shifting the problem around, we would be adding gRPC API surface area just to satisfy this little difference.

> I'm not advocating we have to clean up gRPC right now, necessarily—just that we give it a bit of thought, capture pros/cons here, and see if this is something we can work towards.

I think that can be done external to this proposal.

Reply:
> > I think there is a "middle path" here that we haven't considered, which is to extend the gRPC API to be more RESTful (where needed; seems like it's already close?), make that an explicit design goal of gRPC going forward, and then build the HTTP API on top of that.
> 
> It was never considered it because I don't think altering or adding more gRPC calls is acceptable for this project.

Why would it not be acceptable?

> > seems like it's already close?
> 
> Can you give a concrete example here?

From a cursory glance, Schedules, Namespaces, and Batch Operations all seem to have very CRUD-like gRPC interfaces. I haven't gone through the protobuf in detail, though, so I'm happy to be corrected if there are some hidden details I'm missing.

(And I will point out workflows as a good counter-example—the gRPC API for workflows does not seem to be very RESTful, based on my reading of this proposal alone.)

> > And yes, I'm aware this is just shifting the problem around, code-wise
> 
> Not just shifting the problem around, we would be adding gRPC API surface area just to satisfy this little difference.

Right, we would be. If this is a "little difference", I don't see that it would add very much surface area to gRPC, though? Again, happy to be corrected.

> > I'm not advocating we have to clean up gRPC right now, necessarily—just that we give it a bit of thought, capture pros/cons here, and see if this is something we can work towards.
> 
> I think it's an unreasonable rejected alternative, but then again I thought that about the ones I was forced to add here and any that are "what ifs" beyond the goals. I think if we're going to work towards it, it should be a real proposal not a rejected alternative (I don't think we'll ever work towards it though).
> 
> Having said that, if we need a "Rejected Alternative Approach - More RPC Methods" I can add it.

I know a decision has already been made internally and I'm not going to hold up the train on this any further, but I do think it would be good to briefly document what we decided and why (which I'm still not clear on).

Reply:
> Why would it not be acceptable?

I guess I should have made it clearer in the prelude that it is a goal of the project not to alter or add to the existing API. But to clarify - this project does not want to alter or add to the existing API. I would suggest putting forth an alternative proposal if this is a new goal.

> From a cursory glance, Schedules, Namespaces, and Batch Operations all seem to have very CRUD-like gRPC interfaces. I haven't gone through the protobuf in detail, though, so I'm happy to be corrected if there are some hidden details I'm missing.

Namespaces is rarely used (not even available on cloud) so to alter its API would likely be a waste of effort. Batch doesn't have things like update. Schedules very much does not follow basic CRUD wrt updates. To clarify my request, "can you give a concrete example of proto file alterations you would like to see?"

> If this is a "little difference", I don't see that it would add very much surface area to gRPC, though? 

Can you give a concrete example of proto file alterations you would like to see? It will be much easier if we can discuss practicalities instead of hypotheticals. I think you'll find you'll either be suggesting changes to API that is rarely used anyways, or significant alterations to the commonly used API.

> I do think it would be good to briefly document what we decided and why (which I'm still not clear on).

That was the goal of this document. I assumed I wasn't going to have to document obvious things like "why HTTP?" and "why not change the existing API?" but if I need to, ok.

Reply:
> > Why would it not be acceptable?
> 
> I guess I should have made it clearer in the prelude that it is a goal of the project not to alter or add to the existing API. But to clarify - this project does not want to alter or add to the existing API. I would suggest putting forth an alternative proposal if this is a new goal.

That's been clear from the discussion; I'm asking why it's a non-goal.

To be super clear/explicit: given the decision that was already made internally, I would be happy / my concern here would be addressed if we add one more bullet point under [Non-Goals](https://github.com/temporalio/proposals/pull/79/files#diff-7e381adf57b23564610f84e9ffa150a2cf7a558a0e0c6114eb83c3eff42e7487R14) that says something to the effect that we have decided not to touch gRPC because: time/we'll do it later/we don't think it will ever be necessary because $REASON/whatever the actual reason is.

I think the rest of your comment is moot given that gRPC is out of scope. Happy to discuss further if/when we get to that point, though.

Reply:
> > I think the rest of your comment is moot given that gRPC is out of scope.
> 
> What do you mean by "out of scope"? You're suggesting gRPC alterations correct? Or are you saying "my suggestions are out of scope so justifying them or giving concrete samples is moot"?

I mean that the decision has already been made to exclude any gRPC changes from the scope of this project. So I don't think it's worth either of us spending any more time on this, other than to make sure we document that the decision was made and the reason for that decision.

You and Ryland have already reached an agreement here, and I don't want to get in the way of that (especially as someone who is new to the party).

> I think it's as reasonable for me to ask for concrete examples of API alterations you want to make as it is for you to ask why I don't want to make API alterations. It is much easier to respond to the API alterations you are suggesting if I know what API alterations you are suggesting. Can you provide these so we can discuss them instead of hypotheticals? Then it may be clear the harm of your suggestions are going to vastly outweigh their benefits.

I'd be happy to do that if/when we ever decide we actually want to make gRPC and HTTP align with each other. I don't think it's worth writing a proposal right now given that we have already decided not to pursue it.

Reply:
> I don't think it's worth writing a proposal right now given that we have already decided not to pursue it.

:+1: In the future, it will make discussions like this thread easier. If we're willing to discuss alternatives, I think we should be willing to show exactly what we mean if it's otherwise unclear and/or would help the discussion.